[DAG] add basic plotting on Ray DAGs

lchu-ibm opened the pull request:
<!-- Thank you for your contribution! Please review https://github.com/ray-project/ray/blob/master/CONTRIBUTING.rst before opening a pull request. -->

<!-- Please add a reviewer to the assignee section when you create a PR. If you don't have the access to it, we will shortly find a reviewer and assign them to your PR. -->

## Why are these changes needed?

To add basic plotting feature for Ray DAGs. 

### Exposed API
`ray.experimental.dag.plot(dag: DAGNode, to_file=None)`

### Behavior
1. dump the dag plot (Dot) to file.
2. also render the image whenever possible. E.g. if running in Jupyter notebook, the image will not only be saved, but also rendered in the notebook.
3. when to_file is not set (i.e. None), it will be saved to a tempfile for rendering purpose only.  This is common when users plot DAGs in notebook env to explore the DAG structure without wanting to save it to a file.

### Sample result

**basic DAG with explicit node names**
```
import ray

@ray.remote
def a(*args, **kwargs):
    pass

tmp1 = a.options(name='tmp1').bind()
tmp2 = a.options(name='tmp2').bind()
tmp3 = a.options(name='tmp3').bind(tmp1, tmp2)
tmp4 = a.options(name='tmp4').bind()
tmp5 = a.options(name='tmp5').bind(tmp4)
tmp6 = a.options(name='tmp6').bind()
dag = a.bind(tmp3, tmp5, tmp6)

ray.experimental.dag.plot(dag)
```
<img width="343" alt="image" src="https://user-images.githubusercontent.com/20955448/165413176-e66066a9-8286-46d9-b57a-a2d02501e81c.png">

**basic DAG with auto-generated node names**
```
tmp1 = a.bind()
tmp2 = a.bind()
tmp3 = a.bind(tmp1, tmp2)
tmp4 = a.bind()
tmp5 = a.bind(tmp4)
tmp6 = a.bind()
dag = a.bind(tmp3, tmp5, tmp6)

ray.experimental.dag.plot(dag)
```
<img width="331" alt="image" src="https://user-images.githubusercontent.com/20955448/165413194-5a0a69ea-0b5c-4510-b941-cf3e9fbd75a6.png">

**basic Serve Pipeline DAG**
```
dag = fn_hello.bind()
serve_dag = NoargDriver.bind(dag)
serve_dag = ray_dag_to_serve_dag(serve_dag)
```
<img width="340" alt="image" src="https://user-images.githubusercontent.com/20955448/165368415-2ccca294-9ef2-44e8-96ca-ea0274c06d62.png">

**basic Serve Pipeline DAG with input**
```
with InputNode() as dag_input:
    dag = combine.bind(dag_input[0], dag_input[1], kwargs_output=1)
    serve_dag = DAGDriver.bind(dag, input_schema=json_resolver)
    serve_dag = ray_dag_to_serve_dag(serve_dag)
```
<img width="824" alt="image" src="https://user-images.githubusercontent.com/20955448/165368573-9efbdf77-3f44-4bed-b41c-a246b54770e5.png">

**basic Serve Pipeline DAG with chained function**
```
with InputNode() as dag_input:
    output_1 = func_1.bind(dag_input)
    output_2 = func_2.bind(dag_input)
    serve_dag = combine.bind(output_1, output_2)
    serve_dag = ray_dag_to_serve_dag(serve_dag)
```
<img width="465" alt="image" src="https://user-images.githubusercontent.com/20955448/165369064-0cd48a2e-a454-4ae6-bb44-d55908e6e823.png">

**basic Serve Pipeline DAG with class method**
```
model = Model.bind(2, ratio=0.3)
dag = model.forward.bind(dag_input)
serve_dag = DAGDriver.bind(dag, input_schema=json_resolver)
serve_dag = ray_dag_to_serve_dag(serve_dag)
```
<img width="506" alt="image" src="https://user-images.githubusercontent.com/20955448/165371401-9cc16bb8-4e1c-4163-9105-0bf2f9dd92bd.png">

**Serve Pipeline DAG with class method**
```
with InputNode() as dag_input:
    m1 = Model.bind(1)
    m2 = Model.bind(2)
    m1_output = m1.forward.bind(dag_input[0])
    m2_output = m2.forward.bind(dag_input[1])
    combine_output = combine.bind(m1_output, m2_output, kwargs_output=dag_input[2])
    serve_dag = DAGDriver.bind(combine_output, input_schema=json_resolver)
    serve_dag = ray_dag_to_serve_dag(serve_dag)
```
<img width="994" alt="image" src="https://user-images.githubusercontent.com/20955448/165371892-f3b42666-f95c-48b2-8fc4-708687f9ce81.png">


### Future works that are not included in this PR yet (TODOs in the code)
1. **Support more options for graph**, including rankDir, alignment, shapes, structures, etc. 
- different node shapes for different node types (e.g. use box shape for InputNode while oval for FunctionNode/ClassNode)
- align node positions in the graph based on their "levels" or types. (e.g. left align all InputAttributeNodes to segment input-related nodes from the rest)
These would require building multiple sub-graphs and link them together.

2.  **Support more contents in the graph**, including each node's func_body, arg_and_kwargs, etc.
- currently only node name is displayed in the plot, we will need more information in the future, similar to DL visualization where not only layer name is provided, but also layer info including layer shape, dtypes, etc. 



<!-- Please give a short summary of the change and the problem this solves. -->

## Related issue number

<!-- For example: "Closes #1234" -->

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

jiaodong commented:
@lchu-ibm can you also put up a few tests with `@serve.deployment` decorated DAG ? You can see them in `test_pipeline_dag.py`

jiaodong commented on python/ray/experimental/dag/vis_utils.py:
```diff
@@ -0,0 +1,112 @@
+                _dfs(child_node, res, visited)
+
+    edges = []
+    nodes = []
```
this could be a set to make the dfs visited check O(1) ? Feels like leetcoding again :D

jiaodong commented on python/ray/experimental/dag/vis_utils.py:
```diff
@@ -0,0 +1,112 @@
+
+    # close temp file if needed
+    try:
+        tmp_file.close()
```
this can be wrapped in context manager to automatically close it

lchu-ibm replied:
Agreed, but using a context manager, I feel I need to duplicate the code block of "rendering in jupyter notebook" both within the context manager (when to_file is None) and outside the context manager (when to_file is not None).

jiaodong reviewed (commented):
thank for the quick progress ! this will improve UX a lot.

ericl commented:
Wow, this is super cool. Similar comment as above, have you tried this with Serve topologies? Those include actor creation and method calls which might need to be represented differently from task nodes.

jiaodong commented:
> when duplicated name appears and requires incremental suffix, maybe we should add suffix in certain order (e.g. from downstream to upstream or vise versa), rather than add suffix in a random traversal order.

To clarify I think current traversal order for a DAG is deterministic given it's always done bottom-up and executed by a single threaded python function / pickle. You're feel to try out a super large DAG and see if that behavior holds tho :)

iycheng commented on python/ray/experimental/dag/tests/test_plot.py:
```diff
@@ -0,0 +1,30 @@
+    def a(*args, **kwargs):
+        pass
+
+    tmp1 = a.options(name="tmp1").bind()
```
Could you also write some tests without name? It should work out of the box.

We can use the pkg.func as the name.

lchu-ibm commented:
@ericl @jiaodong @iycheng 

I tested all serve dags in test_pipeline_dag.py and all work fine, yay!  

I will push the changes, as long as the new tests later. But I have already updated all the results above in the **Sample result** portion.

jiaodong commented:
Great to know all these dag patterns worked with visualization attached :) After reading the diagram, I think one future improvement is to have better split / representation for nodes and edges that captures topology as well as data flow ..

![image](https://user-images.githubusercontent.com/6372689/165384735-feb18f3a-0cb1-44b8-b70f-4255b03e1940.png)

In this example the InputNode / InputAttributeNode are essentially data nodes, Model is an instance that being used in computation but they appeared the same way :/ Not needed in this PR for sure.

lchu-ibm commented:
> To clarify I think current traversal order for a DAG is deterministic given it's always done bottom-up and executed by a single threaded python function / pickle. You're free to try out a super large DAG and see if that behavior holds tho :)

@jiaodong This is actually the reason I used List rather than Set, I think the current implementation might be a "randomly deterministic" traversal.  I see the current implementation uses Set for `get-children`-related methods (for both get top level children and all children).  I think the potential problem here is:  as we are storing the Node objects in the set, each time we run the same code, the hash key of the objects will be different.  Therefore, the order within the Set (due to Set's "sorted unordered" nature) can change.

E.g. running the following code multiple times:
```
tmp1 = a.options().bind()
tmp2 = a.options().bind()

s = set()
s.add(tmp1)
s.add(tmp2)
s
```

I believe s can be either (tmp1, tmp2) or (tmp2, tmp1).    Same story with `get_children`, this may cause the DAG looks different if we run the same code again.

I could miss something here, but let me know your thought on this.

jiaodong commented:
> > To clarify I think current traversal order for a DAG is deterministic given it's always done bottom-up and executed by a single threaded python function / pickle. You're free to try out a super large DAG and see if that behavior holds tho :)
> 
> @jiaodong This is actually the reason I used List rather than Set, I think the current implementation might be a "randomly deterministic" traversal. I see the current implementation uses Set for `get-children`-related methods (for both get top level children and all children). I think the potential problem here is: as we are storing the Node objects in the set, each time we run the same code, the hash key of the objects will be different. Therefore, the order within the Set (due to Set's "sorted unordered" nature) can change.
> 
> E.g. running the following code multiple times:
> 
> ```
> tmp1 = a.options().bind()
> tmp2 = a.options().bind()
> 
> s = set()
> s.add(tmp1)
> s.add(tmp2)
> s
> ```
> 
> I believe s can be either (tmp1, tmp2) or (tmp2, tmp1). Same story with `get_children`, this may cause the DAG looks different if we run the same code again.
> 
> I could miss something here, but let me know your thought on this.

Ah you're right. We used set to reduce compute cost on critical path but do not guarantee deterministic order in traversal & replacement.

lchu-ibm commented on python/ray/experimental/dag/vis_utils.py:
```diff
@@ -0,0 +1,112 @@
+    # TODO(lchu): For now, we borrow name generator from
+    # Serve, however, we should eventually have a similar
+    # low-level name generator for DAGNode
+    name_generator = DeploymentNameGenerator()
```
sounds great!

I am currently using the following logic to get the name, a little bit enhancement to the existing one in order to cover all types

```
        if isinstance(node, InputNode):
            node_name = "input_node"
        elif isinstance(node, InputAtrributeNode):
            node_name = "input_attributed_node"
        elif isinstance(node, ClassMethodNode):
            node_name = (
                node.get_options().get("name", None) or node._method_name
            )
        else:
            node_name = (
            node.get_options().get("name", None) or node._body.__name__
        )
```

pushing the code right now.

lchu-ibm commented:
@jiaodong 

Ready for another round of review.

Changes:

1. change Set to List in getting children methods to address above discussion.
2. Add tests to test all examples from `test_pipeline_dag.py` on plotting results. 
3. Add tests to test both with-name Nodes and without-name Nodes.  cc. @iycheng 

4. created a `DAGNodeNameGenerator` to generate name for all types of DAGNodes.   (TODO cc @jiaodong: after we review this and once it looks good, we can remove the `DeploymentNameGenerator`)

lchu-ibm commented:
> Great to know all these dag patterns worked with visualization attached :) After reading the diagram, I think one future improvement is to have better split / representation for nodes and edges that captures topology as well as data flow ..
> 
> ![image](https://user-images.githubusercontent.com/6372689/165384735-feb18f3a-0cb1-44b8-b70f-4255b03e1940.png)
> 
> In this example the InputNode / InputAttributeNode are essentially data nodes, Model is an instance that being used in computation but they appeared the same way :/ Not needed in this PR for sure.

Agreed!  That's also on my TODO list re. "alignment" and "shape":

> 1. Support more options for graph, including rankDir, alignment, shapes, structures, etc

jiaodong commented on python/ray/experimental/dag/vis_utils.py:
```diff
@@ -110,3 +108,42 @@ def plot(dag: DAGNode, to_file=None):
+    def get_node_name(self, node: DAGNode):
+
+        if isinstance(node, InputNode):
+            node_name = "input_node"
```
We actually have an assertion in Ray DAG that it should never have two `InputNode`s, but sadly can only be enforced when we do actual traversal of the DAG via .execute(). It could be slight UX improvement if we ever detect the need to provide a different suffix for `InputNode` with another uuid when user tries to visualize intermediate DAG.

lchu-ibm replied:
Thought about this as I did see the requirement for having only one InputNode.  There are other interesting scenarios there: what if I purposely name my function node as "input_node", lol.  so enforcing it on naming level might not be that useful. But I agree something can be done here to improve UX.

jiaodong reviewed (commented):
Name generator looks good to me, and some nits about cleaning up test_plot for deployment patterns as we don't need to build or run it.

lchu-ibm commented on python/ray/experimental/dag/vis_utils.py:
```diff
@@ -110,3 +108,42 @@ def plot(dag: DAGNode, to_file=None):
+        elif isinstance(node, ClassMethodNode):
+            node_name = node.get_options().get("name", None) or node._method_name
+        else:
+            node_name = node.get_options().get("name", None) or node._body.__name__
```
Changed it to:

```
        if isinstance(node, InputNode):
            node_name = "input_node"
        elif isinstance(node, InputAtrributeNode):
            node_name = "input_attributed_node"
        elif isinstance(node, ClassMethodNode):
            node_name = node.get_options().get("name", None) or node._method_name
        elif isinstance(node, (ClassNode, FunctionNode)):
            node_name = node.get_options().get("name", None) or node._body.__name__
        elif isinstance(node, (DeploymentNode, DeploymentMethodNode, DeploymentFunctionNode)):
            node_name = node.get_deployment_name()
        else:
            raise ValueError("get_node_name() should only be called on DAGNode instances.")
```
to cover all types of DAGNode.

And I agree we would need to clean some parts in DeploymentNode as DeploymentNodes are always generated with a name, using the same logic here (which is the logic I borrowed).

lchu-ibm commented:
@jiaodong I think there is a typo in the naming:

`InputAtrributeNode`,  shouldn't it be `InputAttributeNode`?

I can go ahead to make the change unless this is done on purpose.

jiaodong commented:
> @jiaodong I think there is a typo in the naming:
> 
> `InputAtrributeNode`, shouldn't it be `InputAttributeNode`?
> 
> I can go ahead to make the change unless this is done on purpose.

yes it's typo 🤦 please change it

jiaodong commented:
new PR looks good to me now, only input is changing the serve_dag to call

```
with DeploymentNameGenerator() as deployment_name_generator:
    serve_root_dag = ray_dag_root_node.apply_recursive(
        lambda node: transform_ray_dag_to_serve_dag(node, deployment_name_generator)
    )
```

 in test (otherwise it's still ray dag, not serve dag with DeploymentNode types) and move these to serve/tests  folder instead since they're not ray core dag related, then remove DeploymentNameGenerator

lchu-ibm commented:
> new PR looks good to me now, only input is changing the serve_dag to call
> 
> ```
> with DeploymentNameGenerator() as deployment_name_generator:
>     serve_root_dag = ray_dag_root_node.apply_recursive(
>         lambda node: transform_ray_dag_to_serve_dag(node, deployment_name_generator)
>     )
> ```
> 
> in test (otherwise it's still ray dag, not serve dag with DeploymentNode types) and move these to serve/tests folder instead since they're not ray core dag related, then remove DeploymentNameGenerator

Done. serve-related tests are now moved under `serve/tests/test_pipeline_dag_plot.py`, each with `serve_dag = ray_dag_to_serve_dag(serve_dag)` to convert them from Ray DAG to Serve DAG.

I have also removed DeploymentNameGenerator and replace all usages with `ray.experimental.dag.utils.DAGNodeNameGenerator` which generate names on all 7 subclasses of DAGNode.

lchu-ibm commented:
@jiaodong 
- [ ] Reformat + Lint 
- [ ] Rebase
- [ ] local test on all tests in `ray/experimental/dag/tests`

Just pushed the latest changes and let's see CI later.

suquark commented on python/ray/experimental/dag/dag_node.py:
```diff
@@ -84,47 +83,51 @@ def execute(self, *args, **kwargs) -> Union[ray.ObjectRef, ray.actor.ActorHandle
+        children = []
         for a in self.get_args():
             if isinstance(a, DAGNode):
-                children.add(a)
```
Since python3.6, `set` is ordered. New duplicated objects would not change the order in the set. Why replace it with a list?

lchu-ibm replied:
@suquark for the same objects, yes. But consider the following case:
```
node1 = f.bind()
node2 = f.bind()

s = set()
s.add(node1)
s.add(node2)
```

running above multiple times, you will get s as either (node1, node2) or (node2, node1), as the hash key of the object changed thus the order of the objects are not guaranteed with same code re-run.

suquark replied:
thanks for the answer! just noticed that this has been answered above. maybe add some comments in the code explaining that?

lchu-ibm replied:
@suquark sure! will do!

jiaodong replied:
@suquark from the link i saw a comment "@user474491 Unlike dict, set in Python 3.7+ unfortunately does not preserve order" does it mean CPython 3.6+ supports ordered `dict()` but not for `set()` ?

lchu-ibm replied:
@jiaodong with your comment I just realized that I misunderstood @suquark 's initial post.  And I think @suquark is right if we have **Insertion Ordered** set.  But I think the set implementation is not yet there as you pointed out.

jiaodong replied:
The good news here is the function this file touched wasn't on critical path of Ray DAG execution, only if we want to introspect its structure :) But in future development I will keep this topic in mind -- I think have efficient but also ordering guarantee is favorable on the execution path of Ray DAG, it also makes traversal output order deterministic .

jiaodong commented on python/ray/experimental/dag/py_obj_scanner.py:
```diff
@@ -45,7 +45,7 @@ def __init__(self):
+        from ray.experimental.dag.input_node import InputNode, InputAttributeNode
         from ray.serve.pipeline.deployment_node import DeploymentNode
         from ray.serve.pipeline.deployment_method_node import DeploymentMethodNode
         from ray.serve.pipeline.deployment_function_node import DeploymentFunctionNode
```
cc: @ericl we're seeing some CI failures due to importing serve modules here implies ray core user requires serve dependency such as fastapi. Do you have anything in mind that we can remove these local imports and serve import for pickling ?

lchu-ibm replied:
To add a little bit on this: CI won't complain much here as we did a local import.  But during usage, this might be bothering to users, e.g. 
```
@ray.remote
def a(*args, **kwargs):
    pass

a.bind().get_children()
```
will throw errors like `ModuleNotFoundError: No module named 'fastapi'. You can run `pip install "ray[serve]"` to install all Ray Serve dependencies.`

jiaodong commented:
CI is green now and we're doing branch cut tomorrow. cc: @ericl @iycheng @suquark for another look.

ericl reviewed (approved):
LGTM, but we should add docs in the future. @jiaodong do you own the productionization plan for the dag module?

jiaodong commented:
@ericl that's what i've been doing and will continue :) We should sync some time for next steps of DAG in Ray 2.0
